consensus/src/consensus/inner.rs

```diff
@@ -0,0 +1,508 @@
+            // no blocks
+            let hash = self.public.genesis_block.header.hash();
+            let block = self.public.genesis_block.clone();
+            self.storage.insert_block(&block).await?;
+            self.commit_block(&hash, &block).await?;
+        }
+        self.catch_up().await?;
+        Ok(())
+    }
+
+    pub async fn agent(mut self, mut receiver: mpsc::Receiver<ConsensusMessage>) {
+        self.init()
+            .await
+            .expect("failed to initialize ledger & storage with genesis block");
+
+        while let Some(message) = receiver.recv().await {
+            match message {
+                ConsensusMessage::ReceiveTransaction(transaction, response) => {
+                    response.send(self.receive_transaction(transaction)).ok();
+                }
+                ConsensusMessage::VerifyTransaction(transaction, response) => {
+                    let out = match self.verify_transaction(&*transaction) {
+                        Ok(out) => out,
+                        Err(e) => {
+                            error!(
+                                "failed to validate transaction -- note this does not mean the transaction was valid or invalid: {:?}",
+                                e
+                            );
+                            false
+                        }
+                    };
+                    response.send(out).ok();
+                }
+                ConsensusMessage::ReceiveBlock(block, response) => match self.receive_block(&block).await {
+                    Ok(()) => {
+                        response.send(true).ok();
+                    }
+                    Err(e) => {
+                        error!("failed receiving block: {:?}", e);
+                        response.send(false).ok();
+                    }
+                },
+                ConsensusMessage::FetchMemoryPool(size, response) => {
+                    let out = self
+                        .memory_pool
+                        .get_candidates(size)
+                        .map(|x| x.into_iter().cloned().collect());
+                    response.send(out).ok();
+                }
+                ConsensusMessage::CreateTransaction(request, response) => {
+                    let out = self.create_transaction(*request);
+                    response.send(out).ok();
+                }
+                ConsensusMessage::CreatePartialTransaction(request, response) => {
+                    let out = self.create_partial_transaction(request);
+                    response.send(out).ok();
+                }
+            }
+        }
+    }
+
+    /// Check if the transaction is valid.
+    fn verify_transaction(&self, transaction: &SerialTransaction) -> Result<bool, ConsensusError> {
+        if !self
+            .public
+            .parameters
+            .authorized_inner_snark_ids
+            .contains(&to_bytes![transaction.inner_circuit_id]?)
```

Comment:
`inner_circuit_id` is a `Digest` which is a `SmallVec` internally, so we shouldn't need this intermediate allocation via `to_bytes`